instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java

```diff
@@ -1,23 +1,91 @@
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.kstream.TransformerSupplier;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
 
-/** Use this class to decorate your Kafka Streams client. */
+/** Use this class to decorate Kafka Stream Topologies and enable Tracing. */
 public final class KafkaStreamsTracing {
 
-  public static KafkaStreamsTracing create(KafkaTracing kafkaTracing) {
-    return new KafkaStreamsTracing(kafkaTracing);
+  final Tracing tracing;
+  final TraceContext.Extractor<Headers> extractor;
+
+  KafkaStreamsTracing(Builder builder) { // intentionally hidden constructor
+    this.tracing = builder.tracing;
+    this.extractor = tracing.propagation().extractor(KafkaStreamsPropagation.GETTER);
+  }
+
+  public static KafkaStreamsTracing create(Tracing tracing) {
+    return new KafkaStreamsTracing.Builder(tracing).build();
+  }
+
+  static void addTags(ProcessorContext processorContext, SpanCustomizer result) {
+    result.tag(KafkaStreamsTags.KAFKA_STREAMS_APPLICATION_ID_TAG, processorContext.applicationId());
+    result.tag(KafkaStreamsTags.KAFKA_STREAMS_TASK_ID_TAG, processorContext.taskId().toString());
+  }
+
+  /**
+   * Creates a {@link KafkaStreams} instance with a tracing-enabled {@link KafkaClientSupplier}. All
+   * Topology Sources and Sinks (including internal Topics) will create a Spans on records
+   * processed.
+   */
+  public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
+    return new KafkaStreams(topology, streamsConfig, kafkaClientSupplier());
   }
 
-  final KafkaTracing kafkaTracing;
+  /**
+   * Create a tracing-decorated {@link ProcessorSupplier}
+   */
+  public <K, V> ProcessorSupplier<K, V> processor(String name, Processor<K, V> processor) {
+    return new TracingProcessorSupplier<>(this, name, processor);
+  }
 
-  KafkaStreamsTracing(KafkaTracing kafkaTracing) { // intentionally hidden constructor
-    this.kafkaTracing = kafkaTracing;
+  /**
+   * Create a tracing-decorated {@link TransformerSupplier}
+   */
+  public <K, V, R> TransformerSupplier<K, V, R> transformer(String name,
+      Transformer<K, V, R> transformer) {
+    return new TracingTransformerSupplier<>(this, name, transformer);
   }
 
   /** Returns a client supplier which traces send and receive operations. */
-  public KafkaClientSupplier kafkaClientSupplier() {
+  KafkaClientSupplier kafkaClientSupplier() {
+    KafkaTracing kafkaTracing = KafkaTracing.create(tracing);
     return new TracingKafkaClientSupplier(kafkaTracing);
   }
+
+  Span nextSpan(ProcessorContext context) {
+    TraceContextOrSamplingFlags extracted = extractor.extract(context.headers());
+    Span result = tracing.tracer().nextSpan(extracted);
+    if (!result.isNoop()) {
```

Comment:
is there any case where this would skip the client instrumentation setup? Ex I would expect there to always be headers here. If there isn't then maybe someone skipped a step.

One way to tell is in your micro-integration tests, check that the root span has kind=CONSUMER

Reply:
Client instrumentation could be skipped it by creating a `KafkaStreams` directly without the builder from `KafkaStreamsTracing#kafkaStreams`, where clients are injected. In this case, processors will create independent spans, or potentially linked to a trace injected by a `TracingKafkaProducer` outside the Kafka Streams application. 

I have a test case for this scenario called: `should_create_spans_from_stream_without_tracing_with_tracing_transformer/processor`. I'll also add `kind=CONSUMER` validation on the input and `PRODUCER` on the output.

Reply:
awesome sauce